### PR TITLE
Added codegen for substrate API

### PR DESCRIPTION
**Pull request by ellistarn**
Issue #, if available:

Description of changes:
Precursor PR to the big parallel change. This adds codegen for DeepCopy and moves /apis under /pkg/apis 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: substrate/cmd/delete.go [hunk 1]
```diff
@@ -20,10 +36,16 @@ func init() {
...
 func Delete(cmd *cobra.Command, args []string) {
-	runtime.Must(substrate.NewController(cmd.Context()).Finalize(cmd.Context(), &v1alpha1.Substrate{
-		ObjectMeta: metav1.ObjectMeta{Name: "testvpc"},
+	ctx := cmd.Context()
+	start := time.Now()
+	name := "test"
+	if err := substrate.NewController(ctx).Reconcile(ctx, &v1alpha1.Substrate{
+		ObjectMeta: metav1.ObjectMeta{Name: name, DeletionTimestamp: &metav1.Time{Time: time.Now()}},
 		Spec: v1alpha1.SubstrateSpec{
 			VPC: &v1alpha1.VPCSpec{CIDR: "10.0.0.0/16"},
```
Comment (prateekgogia) [inline substrate/cmd/delete.go:45]: Why do we need the CIDR in delete command?

File: substrate/pkg/apis/v1alpha1/substrate_defaults.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"context"
+
+	"knative.dev/pkg/ptr"
+)
+
+// SetDefaults for the resource
+func (s *Substrate) SetDefaults(ctx context.Context) {
+	if s.Spec.InstanceType == nil {
+		s.Spec.InstanceType = ptr.String("t4g.nano")
```
Comment (prateekgogia) [inline substrate/pkg/apis/v1alpha1/substrate_defaults.go:26]: Do we want to switch to micro here?

File: substrate/pkg/apis/v1alpha1/substrate_validation.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"context"
+
+	"knative.dev/pkg/apis"
+)
+
+func (s *Substrate) Validate(ctx context.Context) (errs *apis.FieldError) {
+	if len(s.Name) == 0 {
+		return errs.Also(apis.ErrMissingField("name"))
+	}
```
Comment (prateekgogia) [inline substrate/pkg/apis/v1alpha1/substrate_validation.go:26]: I think even the name can be optional, we can default to something like `substrate` or `kit-substrate`

**Review by prateekgogia — APPROVED** [review]
Some minor comments! Not blocking!
LGTM!

File: substrate/pkg/apis/v1alpha1/substrate_validation.go (hunk 3 above)
Comment (ellistarn) [reply substrate/pkg/apis/v1alpha1/substrate_validation.go:26]: Agreed. Let's get to this later. This isn't even wired up at the mo.

**Pull request merged**